File: src/application/agent_tools.cpp
#include "application/agent_tools.hpp"

#include <array>
#include <limits>
#include <stdexcept>
#include <utility>

#include "catalogs/generated_catalog.hpp"
#include "query/cxq.hpp"
#include "query/exec.hpp"
#include "util/json_read.hpp"
#include "util/version.hpp"

namespace cidx::agent {
namespace {

constexpr std::array<std::string_view, 2> kTools = {"query", "explain"};

const json_out::Value *field(const json_out::Value &value,
                             std::string_view name) {
  if (value.t != json_out::Value::T::Obj) {
    return nullptr;
  }
  for (const auto &[key, child] : value.o) {
    if (key == name) {
      return &child;
    }
  }
  return nullptr;
}

std::string required_string(const json_out::Value &object,
                            std::string_view name) {
  const json_out::Value *value = field(object, name);
  if (value == nullptr || value->t != json_out::Value::T::Str ||
      value->s.empty()) {
    throw std::invalid_argument("E_PROTOCOL_SCHEMA: '" + std::string(name) +
                                "' must be a non-empty string");
  }
  return value->s;
}

int64_t required_integer(const json_out::Value &object, std::string_view name) {
  const json_out::Value *value = field(object, name);
  if (value == nullptr || value->t != json_out::Value::T::Int) {
    throw std::invalid_argument("E_PROTOCOL_SCHEMA: '" + std::string(name) +
                                "' must be an integer");
  }
  return value->i;
}

protocol::ResultEnvelope envelope_base(const IndexIdentity &index,
                                       std::string operation) {
  protocol::ResultEnvelope result;
  result.operation = std::move(operation);
  result.identity.workspace = index.workspace;
  result.identity.index =
      "semantic-index/schema/" + std::to_string(index.schema_version);
  result.identity.fact_sets = {"symbols"};
  result.identity.freshness = index.freshness;
  result.identity.source_revision = index.source_revision;
  result.identity.source_fingerprint = index.source_fingerprint;
  result.producer.package = "cidx";
  result.producer.version = std::string(version::kFullProductVersion);
  result.producer.backend = "cpp";
  result.producer.schema_version = protocol::kProtocolVersion;
  result.evidence.push_back(
      protocol::EvidenceNode{.id = "queryplan",
                             .evidence_class = "derived",
                             .trust = "producer-verified",
                             .summary = "bounded QueryPlan execution",
                             .source = std::nullopt,
                             .children = {}});
  result.artifacts.push_back(
      protocol::ArtifactRef{.kind = "semantic-index",
                            .id = result.identity.index,
                            .schema_version = index.schema_version,
                            .catalog_version = catalog::kCatalogVersion,
                            .catalog_hash = std::string(catalog::kCatalogHash),
                            .generation = std::nullopt});
  return result;
}

protocol::ResultEnvelope failure(const IndexIdentity &index,
                                 std::string operation, std::string code,
                                 std::string message) {
  protocol::ResultEnvelope result = envelope_base(index, std::move(operation));
  result.status = code == "policy_refuted" ? protocol::Status::Refuted
                                           : protocol::Status::Error;
  result.completeness.state = "unknown";
  result.completeness.stale = index.freshness == "stale";
  result.result = json_out::Value::obj({});
  result.diagnostics.push_back(
      protocol::Diagnostic{.code = std::move(code),
                           .severity = "error",
                           .message = std::move(message),
                           .next_action = std::nullopt});
  return result;
}

void require_row_evidence(protocol::ResultEnvelope &result) {
  if (result.completeness.truncated ||
      result.result.t != json_out::Value::T::Obj) {
    return;
  }
  const json_out::Value *rows = field(result.result, "rows");
  if (rows == nullptr || rows->t != json_out::Value::T::Arr) {
    return;
  }
  for (const auto &row : rows->a) {
    const json_out::Value *file = field(row, "file");
    const json_out::Value *line = field(row, "line");
    if (file == nullptr || file->t == json_out::Value::T::Null ||
        line == nullptr || line->t == json_out::Value::T::Null) {
      result.status = protocol::Status::Unknown;
      result.completeness.state = "unknown";
      result.diagnostics.push_back(protocol::Diagnostic{
          .code = "missing_evidence",
          .severity = "warning",
          .message = "result row has no resolvable file/line provenance",
          .next_action = "select file and line evidence or inspect the index"});
      return;
    }
  }
}

void apply_freshness(protocol::ResultEnvelope &result,
                     const IndexIdentity &index) {
  if (index.freshness == "stale") {
    result.status = protocol::Status::Unknown;
    result.completeness.state = "unknown";
    result.completeness.stale = true;
    result.diagnostics.push_back(protocol::Diagnostic{
        .code = "stale_input",
        .severity = "error",
        .message = "index contents are stale for the workspace",
        .next_action =
            "re-index the affected sources before relying on this result"});
  } else if (index.freshness != "current") {
    result.status = protocol::Status::Unknown;
    result.completeness.state = "unknown";
    result.diagnostics.push_back(protocol::Diagnostic{
        .code = "unknown",
        .severity = "warning",
        .message = "index freshness could not be verified",
        .next_action =
            "stamp or re-index the workspace before relying on this result"});
  } else {
    result.status = protocol::Status::Complete;
    result.completeness.state = "complete";
  }
}

IndexIdentity index_identity(application::ApplicationContext &context) {
  if (context.read_ports().query != nullptr) {
    return context.read_ports().query->index_identity();
  }
  IndexIdentity fallback;
  fallback.schema_version = 1;
  fallback.workspace = "workspace:agent";
  fallback.freshness = "unverifiable";
  return fallback;
}

} // namespace

std::string_view tool_name(Tool tool) noexcept {
  switch (tool) {
  case Tool::query:
    return kTools[0];
  case Tool::explain:
    return kTools[1];
  }
  return {};
}

std::span<const std::string_view> tool_catalog() noexcept { return kTools; }

Request ToolService::decode_request(const std::string &json) {
  const json_out::Value root = json_read::parse(json);
  if (root.t != json_out::Value::T::Obj) {
    throw std::invalid_argument("E_PROTOCOL_SCHEMA: request must be an object");
  }
  const int64_t version = required_integer(root, "version");
  if (version != kProtocolVersion) {
    throw std::invalid_argument(
        "E_PROTOCOL_VERSION: unsupported agent protocol version " +
        std::to_string(version));
  }
  const std::string name = required_string(root, "tool");
  Request request;
  request.version = static_cast<int>(version);
  if (name == "query") {
    request.tool = Tool::query;
  } else if (name == "explain") {
    request.tool = Tool::explain;
  } else {
    throw std::invalid_argument("E_TOOL: unsupported agent tool '" + name +
                                "'");
  }
  request.query.expression = required_string(root, "cxq");
  request.query.output = application::QueryOutput::json;
  request.query.explain = request.tool == Tool::explain;
  if (const json_out::Value *budget = field(root, "budget");
      budget != nullptr) {
    if (budget->t != json_out::Value::T::Obj) {
      throw std::invalid_argument(
          "E_PROTOCOL_SCHEMA: 'budget' must be an object");
    }
    request.budget.max_results = required_integer(*budget, "max_results");
  }
  if (request.budget.max_results < 1 ||
      request.budget.max_results > kMaxResults) {
    throw std::invalid_argument("E_BUDGET: max_results must be between 1 and " +
                                std::to_string(kMaxResults));
  }
  request.query.max_results = request.budget.max_results;
  return request;
}

protocol::ResultEnvelope
ToolService::invoke(const Request &request,
                    application::ApplicationContext &context) {
  const IndexIdentity index = index_identity(context);
  const std::string operation(tool_name(request.tool));
  if (request.version != kProtocolVersion) {
    return failure(index, operation, "invalid_input",
                   "E_PROTOCOL_VERSION: unsupported agent protocol version " +
                       std::to_string(request.version));
  }
  if (tool_name(request.tool).empty()) {
    return failure(index, "agent", "invalid_input",
                   "E_TOOL: unsupported agent tool");
  }
  if (request.budget.max_results < 1 ||
      request.budget.max_results > kMaxResults) {
    return failure(index, operation, "invalid_input",
                   "E_BUDGET: max_results must be between 1 and " +
                       std::to_string(kMaxResults));
  }
  if (!context.permits(
          application::capability_bit(application::Capability::index_read))) {
    return failure(index, operation, "policy_refuted",
                   "agent tools require the index_read capability");
  }
  if (context.read_ports().query == nullptr) {
    return failure(index, operation, "backend_error",
                   "query read port is not installed");
  }
  try {
    const query::Plan plan = query::parse_cxq(request.query.expression);
    query::Executor executor(*context.read_ports().query);
    if (request.tool == Tool::query) {
      query::Result executed =
          executor.run(plan, std::nullopt, request.budget.max_results);
      protocol::ResultEnvelope result = executed.to_envelope();
      result.operation = operation;
      require_row_evidence(result);
      return result;
    }
    protocol::ResultEnvelope result = envelope_base(index, operation);
    result.result = executor.explain(plan);
    apply_freshness(result, index);
    return result;
  } catch (const std::exception &error) {
    return failure(index, operation, "invalid_input", error.what());
  }
}

json_out::Value
ToolService::encode_response(const Request &request,
                             const protocol::ResultEnvelope &response) {
  return json_out::Value::obj(
      {{"protocol", json_out::Value::of(std::string(kProtocol))},
       {"version", json_out::Value::of(kProtocolVersion)},
       {"tool", json_out::Value::of(std::string(tool_name(request.tool)))},
       {"response", response.to_json()},
       {"truncated", json_out::Value::of(response.completeness.truncated)},
       {"budget",
        json_out::Value::obj(
            {{"max_results", json_out::Value::of(request.budget.max_results)},
             {"exhausted",
              json_out::Value::of(response.completeness.budget.has_value())},
             {"exhausted_at",
              response.completeness.budget
                  ? json_out::Value::of(*response.completeness.budget)
                  : json_out::Value::null()}})}});
}

json_out::Value
ToolService::invoke_json(const std::string &json,
                         application::ApplicationContext &context) {
  try {
    const Request request = decode_request(json);
    const protocol::ResultEnvelope response = invoke(request, context);
    return encode_response(request, response);
  } catch (const std::exception &error) {
    Request request;
    request.tool = Tool::query;
    request.query.expression = "";
    const protocol::ResultEnvelope response = failure(
        index_identity(context), "query", "invalid_input", error.what());
    return encode_response(request, response);
  }
}

} // namespace cidx::agent

File: python/indexer/agent_tools.py
"""Versioned read-only ``query`` and ``explain`` agent tools.

The adapter is deliberately limited to the maintained storage/query surface:
it opens SQLite with ``mode=ro`` and ``PRAGMA query_only`` and delegates CXQ
parsing, normalization, execution, and result-envelope construction to the
existing QueryPlan implementation.
"""

from __future__ import annotations

from dataclasses import dataclass
import argparse
import json
from pathlib import Path
import sqlite3
import os
import sys
from typing import Any
from urllib.parse import quote

from .generated_catalog import CATALOG_HASH, CATALOG_VERSION
from .queryplan import Executor, PlanError, parse_cxq
from .result_protocol import (
    ArtifactRef,
    Completeness,
    Diagnostic,
    Evidence,
    Identity,
    Producer,
    ResultEnvelope,
    Status,
)
from .storage import Storage

PROTOCOL = "cidx.agent/v1"
PROTOCOL_VERSION = 1
MAX_RESULTS = 10_000
TOOLS = ("query", "explain")


@dataclass(frozen=True)
class AgentRequest:
    version: int
    tool: str
    cxq: str
    max_results: int = 1000

    def __post_init__(self) -> None:
        if type(self.version) is not int:
            raise ValueError("E_PROTOCOL_SCHEMA: 'version' must be an integer")
        if self.version != PROTOCOL_VERSION:
            raise ValueError(
                f"E_PROTOCOL_VERSION: unsupported agent protocol version {self.version}"
            )
        if self.tool not in TOOLS:
            raise ValueError(f"E_TOOL: unsupported agent tool {self.tool!r}")
        if not isinstance(self.cxq, str) or not self.cxq:
            raise ValueError("E_PROTOCOL_SCHEMA: 'cxq' must be a non-empty string")
        if type(self.max_results) is not int or not 1 <= self.max_results <= MAX_RESULTS:
            raise ValueError(
                f"E_BUDGET: max_results must be between 1 and {MAX_RESULTS}"
            )

    @classmethod
    def from_dict(cls, value: dict[str, Any]) -> "AgentRequest":
        version = value.get("version")
        if type(version) is not int:
            raise ValueError("E_PROTOCOL_SCHEMA: 'version' must be an integer")
        if version != PROTOCOL_VERSION:
            raise ValueError(
                f"E_PROTOCOL_VERSION: unsupported agent protocol version {version}"
            )
        tool = value.get("tool")
        if tool not in TOOLS:
            raise ValueError(f"E_TOOL: unsupported agent tool {tool!r}")
        cxq = value.get("cxq")
        if not isinstance(cxq, str) or not cxq:
            raise ValueError("E_PROTOCOL_SCHEMA: 'cxq' must be a non-empty string")
        budget = value.get("budget", {})
        if not isinstance(budget, dict):
            raise ValueError("E_PROTOCOL_SCHEMA: 'budget' must be an object")
        max_results = budget.get("max_results", 1000)
        if type(max_results) is not int or not 1 <= max_results <= MAX_RESULTS:
            raise ValueError(
                f"E_BUDGET: max_results must be between 1 and {MAX_RESULTS}"
            )
        return cls(version, tool, cxq, max_results)


def _envelope_identity(index: Any) -> Identity:
    return Identity(
        workspace=index.workspace,
        index=f"semantic-index/schema/{index.schema_version}",
        fact_sets=("symbols",),
        freshness=index.freshness,
        source_revision=index.source_revision,
        source_fingerprint=index.source_fingerprint,
    )


def _explain_envelope(explanation: dict[str, Any], index: Any) -> ResultEnvelope:
    freshness = index.freshness
    if freshness == "stale":
        status = Status.UNKNOWN
        state = "unknown"
        diagnostics = [
            Diagnostic(
                "stale_input",
                "error",
                "index contents are stale for the workspace",
                "re-index the affected sources before relying on this result",
            )
        ]
    elif freshness != "current":
        status = Status.UNKNOWN
        state = "unknown"
        diagnostics = [
            Diagnostic(
                "unknown",
                "warning",
                "index freshness could not be verified",
                "stamp or re-index the workspace before relying on this result",
            )
        ]
    else:
        status = Status.COMPLETE
        state = "complete"
        diagnostics = []
    return ResultEnvelope(
        operation="explain",
        status=status,
        identity=_envelope_identity(index),
        producer=Producer(backend="python"),
        completeness=Completeness(state=state, stale=freshness == "stale"),
        result=explanation,
        diagnostics=diagnostics,
        evidence=[Evidence("queryplan", "derived", "producer-verified", "bounded QueryPlan execution")],
        artifacts=[
            ArtifactRef(
                "semantic-index",
                f"semantic-index/schema/{index.schema_version}",
                index.schema_version,
                CATALOG_VERSION,
                CATALOG_HASH,
            )
        ],
    )


class AgentTools:
    """The complete agent catalog and its read-only implementation."""

    def __init__(self, db: Storage, connection: sqlite3.Connection | None = None):
        self._db = db
        self._connection = connection

    @classmethod
    def from_path(cls, path: str | Path) -> "AgentTools":
        absolute = str(Path(path).expanduser().resolve())
        uri = f"file:{quote(absolute, safe='/')}?mode=ro"
        connection = sqlite3.connect(uri, uri=True)
        connection.row_factory = sqlite3.Row
        connection.execute("PRAGMA query_only = ON")
        return cls(Storage.from_connection(connection, absolute), connection)

    @classmethod
    def catalog(cls) -> tuple[str, ...]:
        return TOOLS

    def close(self) -> None:
        if self._connection is not None:
            self._connection.close()
            self._connection = None

    def __enter__(self) -> "AgentTools":
        return self

    def __exit__(self, *_: object) -> None:
        self.close()

    def invoke(self, request: AgentRequest | dict[str, Any]) -> dict[str, Any]:
        if isinstance(request, dict):
            request = AgentRequest.from_dict(request)
        if request.version != PROTOCOL_VERSION:
            raise ValueError(
                f"E_PROTOCOL_VERSION: unsupported agent protocol version {request.version}"
            )
        plan = parse_cxq(request.cxq)
        executor = Executor(self._db)
        if request.tool == "query":
            envelope = executor.run(plan, result_cap=request.max_results).to_envelope_dict()
            rows = envelope["result"].get("rows", [])
            if any(
                row.get("file") is None or row.get("line") is None for row in rows
            ):
                if not envelope["completeness"]["truncated"]:
                    envelope["status"] = "unknown"
                    envelope["completeness"]["state"] = "unknown"
                envelope["diagnostics"].append(
                    {
                        "code": "missing_evidence",
                        "severity": "warning",
                        "message": "result row has no resolvable file/line provenance",
                        "next_action": "select file and line evidence or inspect the index",
                    }
                )
        else:
            envelope = _explain_envelope(
                executor.explain(plan), self._db.index_identity()
            ).to_dict()
        exhausted = envelope["completeness"]["budget"]
        return {
            "protocol": PROTOCOL,
            "version": PROTOCOL_VERSION,
            "tool": request.tool,
            "response": envelope,
            "truncated": envelope["completeness"]["truncated"],
            "budget": {
                "max_results": request.max_results,
                "exhausted": exhausted is not None,
                "exhausted_at": exhausted,
            },
        }

    def invoke_json(self, line: str) -> str:
        try:
            value = json.loads(line)
            if not isinstance(value, dict):
                raise ValueError("E_PROTOCOL_SCHEMA: request must be an object")
            return json.dumps(
                self.invoke(value), ensure_ascii=True, separators=(",", ":")
            )
        except (PlanError, ValueError, json.JSONDecodeError) as error:
            index = self._db.index_identity()
            envelope = ResultEnvelope(
                operation="query",
                status=Status.ERROR,
                identity=_envelope_identity(index),
                producer=Producer(backend="cpp"),
                completeness=Completeness(
                    state="unknown", stale=index.freshness == "stale"
                ),
                result={},
                diagnostics=[Diagnostic("invalid_input", "error", str(error))],
                evidence=[Evidence("queryplan", "derived", "producer-verified", "bounded QueryPlan execution")],
                artifacts=[
                    ArtifactRef(
                        "semantic-index",
                        f"semantic-index/schema/{index.schema_version}",
                        index.schema_version,
                        CATALOG_VERSION,
                        CATALOG_HASH,
                    )
                ],
            ).to_dict()
            return json.dumps(
                {
                    "protocol": PROTOCOL,
                    "version": PROTOCOL_VERSION,
                    "tool": "query",
                    "response": envelope,
                    "truncated": False,
                    "budget": {
                        "max_results": 1000,
                        "exhausted": False,
                        "exhausted_at": None,
                    },
                },
                ensure_ascii=True,
                separators=(",", ":"),
            )


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="cidx agent NDJSON adapter")
    parser.add_argument(
        "--index",
        default=str(Path(os.environ.get("INDEXER_CACHE", "~/.cache/cidx")) / "index.db"),
    )
    args = parser.parse_args(argv)
    with AgentTools.from_path(args.index) as tools:
        for line in sys.stdin:
            if line.strip():
                print(tools.invoke_json(line), flush=True)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())


__all__ = [
    "AgentRequest",
    "AgentTools",
    "MAX_RESULTS",
    "PROTOCOL",
    "PROTOCOL_VERSION",
    "TOOLS",
    "main",
]
